src/data/list/palindrome.lean

```diff
@@ -0,0 +1,65 @@
+/-
+Copyright (c) 2020 Google LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Wong
+-/
+
+import data.list.basic
+open list
+
+/-!
+# Palindromes
+
+This module defines *palindromes*, lists which are equal to their reverse.
+
+The main result is the `palindrome` inductive type, and its associated `palindrome.rec_on` induction
+principle. Also provided are conversions to and from other equivalent definitions.
+
+## References
+
+* [Pierre Castéran, *On palindromes*][casteran]
+
+[casteran]: https://www.labri.fr/perso/casteran/CoqArt/inductive-prop-chap/palindrome.html
+
+## Tags
+
+palindrome, reverse, induction
+-/
+
+variables {α : Type*}
+
+/--
+`palindrome l` asserts that `l` is a palindrome. This is defined inductively:
+
+* The empty list is a palindrome;
+* A list with one element is a palindrome;
+* Adding the same element to both ends of a palindrome results in a bigger palindrome.
+-/
+inductive palindrome : list α → Prop
+| nil : palindrome []
+| singleton : ∀ x, palindrome [x]
+| cons_concat : ∀ x {l}, palindrome l → palindrome (x :: (l ++ [x]))
+
+namespace palindrome
+
+lemma to_reverse_eq {l : list α} (p : palindrome l) : reverse l = l :=
+palindrome.rec_on p rfl (λ _, rfl) (λ x l p h, by simp [h])
+
+lemma of_reverse_eq {l : list α} (h : reverse l = l) : palindrome l :=
+begin
+  suffices : reverse l = l → palindrome l, from this h,
```

Comment:
Is this just `revert h`?

Reply:
Good point. I'll move the hypothesis to the right of the colon for now, as it's shorter.